### ISSUE TITLE
Spawning of Notebook does not have permission to write to persistent volume

### ISSUE THREAD

**Issue by kavemun** [kubeflow/kubeflow#3487]
I am trying to spawn a notebook with workspace volume from the kubeflow UI. However I am running into permission denied errors.

This is quite similar to 
https://github.com/kubeflow/kubeflow/issues/1241
however, changing the permission on the folders does not resolve the problem

```
$ kubectl -n kubeflow get pods
NAME                                                        READY   STATUS             RESTARTS   AGE
ambassador-784d4d757-4fmp8                                  1/1     Running            0          23h
ambassador-784d4d757-r59jm                                  1/1     Running            2          2d19h
ambassador-784d4d757-sxq7k                                  1/1     Running            2          2d19h
argo-ui-77d64677b9-94867                                    1/1     Running            0          23h
centraldashboard-dd96bb455-8wc8b                            1/1     Running            0          2d19h
jupyter-web-app-6f946577d5-d7p84                            1/1     Running            0          2d19h
katib-ui-54b8484d4b-lxhzn                                   1/1     Running            0          2d19h
km-test-notebook-01-0                                       1/1     Running            0          109m
metacontroller-0                                            1/1     Running            0          3d1h
minio-5f6b9b5b7-vnrxt                                       1/1     Running            0          2d19h
ml-pipeline-79474f7b99-ff67x                                1/1     Running            0          23h
ml-pipeline-persistenceagent-c56c89b98-chrbp                1/1     Running            1          23h
ml-pipeline-scheduledworkflow-855c759488-t84x2              1/1     Running            0          2d19h
ml-pipeline-ui-5bd6b6dd4-j67zd                              1/1     Running            0          23h
ml-pipeline-viewer-controller-deployment-5b997b657d-9gw2h   1/1     Running            0          2d19h
mysql-657f87857d-cr54m                                      1/1     Running            0          2d19h
notebook-with-pvc-0                                         0/1     CrashLoopBackOff   13         46m
notebooks-controller-794cd5fdd9-pttvm                       1/1     Running            0          23h
pytorch-operator-86d9cdbcd5-bjf9h                           1/1     Running            177        2d19h
studyjob-controller-6965c8c85-9fppl                         1/1     Running            0          2d19h
tensorboard-5cd9974897-8bzqk                                1/1     Running            0          23h
tf-job-dashboard-bfbb84f87-hvkwq                            1/1     Running            0          23h
tf-job-operator-645ff4f66d-sjk5x                            1/1     Running            46         23h
vizier-core-5c654c5666-s4q9d                                1/1     Running            1          2d19h
vizier-core-rest-7b9d769b9d-pjlh8                           1/1     Running            0          23h
vizier-db-68f977d674-7nv2x                                  1/1     Running            0          3d1h
vizier-suggestion-bayesianoptimization-867fdb7fb8-jrwq6     1/1     Running            0          2d19h
vizier-suggestion-grid-fd5848d7d-g29r7                      1/1     Running            0          2d19h
vizier-suggestion-hyperband-756db8787d-9kmdg                1/1     Running            0          2d19h
vizier-suggestion-random-675f7bfbc6-wl5m6                   1/1     Running            0          2d19h
workflow-controller-5b8d97cc6-qkjg6                         1/1     Running            0          2d19h
```

logs from notebook-with-pvc-0 : 
```
$ kubectl -n kubeflow logs notebook-with-pvc-0
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 528, in get
    value = obj._trait_values[self.name]
KeyError: 'runtime_dir'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/jupyter-notebook", line 10, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.6/site-packages/jupyter_core/application.py", line 266, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/traitlets/config/application.py", line 657, in launch_instance
    app.initialize(argv)
  File "</opt/conda/lib/python3.6/site-packages/decorator.py:decorator-gen-7>", line 2, in initialize
  File "/opt/conda/lib/python3.6/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/notebook/notebookapp.py", line 1628, in initialize
    self.init_configurables()
  File "/opt/conda/lib/python3.6/site-packages/notebook/notebookapp.py", line 1319, in init_configurables
    connection_dir=self.runtime_dir,
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 556, in __get__
    return self.get(obj, cls)
  File "/opt/conda/lib/python3.6/site-packages/traitlets/traitlets.py", line 535, in get
    value = self._validate(obj, dynamic_default())
  File "/opt/conda/lib/python3.6/site-packages/jupyter_core/application.py", line 99, in _runtime_dir_default
    ensure_dir_exists(rd, mode=0o700)
  File "/opt/conda/lib/python3.6/site-packages/jupyter_core/utils/__init__.py", line 13, in ensure_dir_exists
    os.makedirs(path, mode=mode)
  File "/opt/conda/lib/python3.6/os.py", line 210, in makedirs
    makedirs(head, mode, exist_ok)
  File "/opt/conda/lib/python3.6/os.py", line 210, in makedirs
    makedirs(head, mode, exist_ok)
  File "/opt/conda/lib/python3.6/os.py", line 210, in makedirs
    makedirs(head, mode, exist_ok)
  File "/opt/conda/lib/python3.6/os.py", line 220, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/home/jovyan/.local'
```

**Comment by jlewi** [comment]
Can you describe your setup please?

* How are you running Kubernetes?
* What are you using as your default storage class? 
   *  How is it configured with respect to permissions?

The Jupyter UI allows you to select the PVC you want to use. So one option would be to not use the default storage class but to manually create the PVC outside of the UI and set the permissions (e.g. READ/WRITE all) so that Jovyan will be able to read and write to it.

**Comment by kavemun** [comment]
Hi @jlewi,


Kubernetes version 1.14.1
running kube on a cluster with 3 master nodes and 4 workers.
I am using cephfs as storageclass
```
$ kubectl get storageclass
NAME               PROVISIONER       AGE
cephfs (default)   ceph.com/cephfs   15d
```

```
$ kubectl describe storageclass
Name:                  cephfs
IsDefaultClass:        Yes
Annotations:           storageclass.kubernetes.io/is-default-class=true
Provisioner:           ceph.com/cephfs
Parameters:            adminId=admin,adminSecretName=ceph-secret,adminSecretNamespace=kube-system,claimRoot=/kubestorage,monitors=10.60.1.34:6789,10.60.1.35:6789,10.60.1.36:6789
AllowVolumeExpansion:  <unset>
MountOptions:          <none>
ReclaimPolicy:         Delete
VolumeBindingMode:     Immediate
Events:                <none>

```

```
$ kubectl -n kubeflow describe pv pvc-2f2397de-919b-11e9-a44d-525400067096
Name:            pvc-2f2397de-919b-11e9-a44d-525400067096
Labels:          <none>
Annotations:     cephFSProvisionerIdentity: cephfs-provisioner-1
                 cephShare: kubernetes-dynamic-pvc-2f520972-919b-11e9-85c1-f270283a6b3f
                 pv.kubernetes.io/provisioned-by: ceph.com/cephfs
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    cephfs
Status:          Bound
Claim:           kubeflow/kf-workspace-volume
Reclaim Policy:  Delete
Access Modes:    RWO
VolumeMode:      Filesystem
Capacity:        10Gi
Node Affinity:   <none>
Message:         
Source:
    Type:        CephFS (a CephFS mount on the host that shares a pod's lifetime)
    Monitors:    [10.122.1.12:6789 10.122.1.13:6789 10.122.1.14:6789]
    Path:        /kubestorage/kubernetes/kubernetes-dynamic-pvc-2f520972-919b-11e9-85c1-f270283a6b3f
    User:        kubernetes-dynamic-user-2f5209e1-919b-11e9-85c1-f270283a6b3f
    SecretFile:  
    SecretRef:   &SecretReference{Name:ceph-kubernetes-dynamic-user-2f5209e1-919b-11e9-85c1-f270283a6b3f-secret,Namespace:kubeflow,}
    ReadOnly:    false
Events:          <none>
```

```
$ kubectl -n kubeflow describe pvc kf-workspace-volume
Name:          kf-workspace-volume
Namespace:     kubeflow
StorageClass:  cephfs
Status:        Bound
Volume:        pvc-2f2397de-919b-11e9-a44d-525400067096
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: ceph.com/cephfs
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Events:
  Type       Reason                 Age   From                                                                                      Message
  ----       ------                 ----  ----                                                                                      -------
  Normal     Provisioning           19h   ceph.com/cephfs_cephfs-provisioner-59494fb97b-sz846_fb5d6bf5-9179-11e9-85c1-f270283a6b3f  External provisioner is provisioning volume for claim "kubeflow/kf-workspace-volume"
  Normal     ProvisioningSucceeded  19h   ceph.com/cephfs_cephfs-provisioner-59494fb97b-sz846_fb5d6bf5-9179-11e9-85c1-f270283a6b3f  Successfully provisioned volume pvc-2f2397de-919b-11e9-a44d-525400067096
  Normal     ExternalProvisioning   19h   persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "ceph.com/cephfs" or manually created by system administrator
Mounted By:  <none>

```
It might be something to do with the pv setup, I'll try your suggestion meanwhile. Thanks

**Comment by imicer** [comment]
@kavemun Have you resolved your issue? I have met with the same issue with you. If you have resolved it, Could you help to tell me the detail for workaround? Thanks.

**Comment by imicer** [comment]
Additionally, for my case, I used kind tools to create k8s cluster, my nodes run as docker env, I am not sure if this is related to my issue or not.

**Comment by kavemun** [comment]
Hi @imicer, I am currently overwriting the permissions for the workspace folder as @jlewi suggested, it works but I am not very sure what is responsible for setting the permission for the folders yet. I am getting my kube admin to trouble shoot on the cluster as well, but as far as we know, the folders are created with root and permission 755, which the container has no permission to write into which causes the problem.

**Comment by imicer** [comment]
Hi @kavemun 
Thanks for your reply. I am studying kubernetes. I am not very familiar with it. Could you help to provide the detailed steps. I still can't get point for your workaround. 
In kubeflow UI, I created nb02-0 for jupyter, and there is no change after changing  permissions for /home/jovyan in docker for nb02 pod.

My steps are like below:
 `
# find docker id for notebook pods
kubectl -n kubeflow describe pod nb02-0
Name:               nb02-0
Namespace:          kubeflow
Priority:           0
PriorityClassName:  <none>
Node:               moelove-ha-worker3/172.17.0.4
Start Time:         Sat, 22 Jun 2019 22:34:00 +0800
Labels:             controller-revision-hash=nb02-b4555658
                    gcp-cred-secret=user-gcp-sa
                    gcp-cred-secret-filename=user-gcp-sa.json
                    notebook-name=nb02
                    statefulset=nb02
                    statefulset.kubernetes.io/pod-name=nb02-0
Annotations:        <none>
Status:             Running
IP:                 10.38.0.9
Controlled By:      StatefulSet/nb02
Containers:
  nb02:
    Container ID:   docker://dde0e3ff6dffb0d6a51861c71dd63d172e1da1186e7619f271561445830d40fc
    Image:          gcr.io/kubeflow-images-public/tensorflow-1.13.1-notebook-cpu:v0.5.0
    Image ID:       docker-pullable://gcr.io/kubeflow-images-public/tensorflow-1.13.1-notebook-cpu@sha256:5aaccf0267f085afd976342a8e943a9c6cefccef5b554df4e15fa7bf15cbd7a3
    Port:           8888/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Sat, 22 Jun 2019 23:05:27 +0800
      Finished:     Sat, 22 Jun 2019 23:05:27 +0800
    Ready:          False
    Restart Count:  11
    Requests:
      cpu:     500m
      memory:  1Gi
    Environment:
      NB_PREFIX:  /notebook/kubeflow/nb02
    Mounts:
      /home/jovyan from nb02 (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from jupyter-notebook-token-9x44q (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  nb02:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  nb02
    ReadOnly:   false
  jupyter-notebook-token-9x44q:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  jupyter-notebook-token-9x44q
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                    From                         Message
  ----     ------     ----                   ----                         -------
  Normal   Scheduled  34m                    default-scheduler            Successfully assigned kubeflow/nb02-0 to moelove-ha-worker3
  Normal   Pulled     33m (x5 over 34m)      kubelet, moelove-ha-worker3  Container image "gcr.io/kubeflow-images-public/tensorflow-1.13.1-notebook-cpu:v0.5.0" already present on machine
  Normal   Created    33m (x5 over 34m)      kubelet, moelove-ha-worker3  Created container
  Normal   Started    33m (x5 over 34m)      kubelet, moelove-ha-worker3  Started container
  Warning  BackOff    4m55s (x138 over 34m)  kubelet, moelove-ha-worker3  Back-off restarting failed container`
`
# Go into the docker env for notebook pods and change permissions
docker exec -ti moelove-ha-worker3 /bin/bash
root@moelove-ha-worker3:/home# chmod 777 /home/jovyan/
`

**Comment by imicer** [comment]
Hi @kavemun,
Thanks for your reply. I am studying kubernetes. I am not very familiar with it. 
Could you help to provide the detailed steps. I still can't get point for your workaround.
In kubeflow UI, I created nb02-0 for jupyter, and there is no change after changing permissions for /home/jovyan in docker for nb02 pod.
My steps are like below:
1, Find docker id for notebook 
```
kubectl -n kubeflow describe pod nb02-0
Name:               nb02-0
Namespace:          kubeflow
Priority:           0
PriorityClassName:  <none>
Node:               moelove-ha-worker3/172.17.0.4
Start Time:         Sat, 22 Jun 2019 22:34:00 +0800
........................
status:             Running
IP:                 10.38.0.9
Controlled By:      StatefulSet/nb02
Containers:
  nb02:
    Container ID:   docker://ffee1756ec2807717369804f335099a36ec61eb90fffe5a7495fb9d37ba874f6
    Image:          gcr.io/kubeflow-images-public/tensorflow-1.13.1-notebook-cpu:v0.5.0
    Image ID:       docker-pullable://gcr.io/kubeflow-images-public/tensorflow-1.13.1-notebook-cpu@sha256:5aaccf0267f085afd976342a8e943a9c6cefccef5b554df4e15fa7bf15cbd7a3
    Port:           8888/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Sat, 22 Jun 2019 23:30:55 +0800
      Finished:     Sat, 22 Jun 2019 23:30:55 +0800
    Ready:          False
    Restart Count:  16
    Requests:
      cpu:     500m
      memory:  1Gi
    Environment:
      NB_PREFIX:  /notebook/kubeflow/nb02
    Mounts:
      /home/jovyan from nb02 (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from jupyter-notebook-token-9x44q (ro)
Conditions:
.................
Events:
  Type     Reason     Age                    From                         Message
  ----     ------     ----                   ----                         -------
  Normal   Scheduled  57m                    default-scheduler            Successfully assigned kubeflow/nb02-0 to moelove-ha-worker3
  Normal   Pulled     56m (x5 over 57m)      kubelet, moelove-ha-worker3  Container image "gcr.io/kubeflow-images-public/tensorflow-1.13.1-notebook-cpu:v0.5.0" already present on machine
  Normal   Created    56m (x5 over 57m)      kubelet, moelove-ha-worker3  Created container
  Normal   Started    56m (x5 over 57m)      kubelet, moelove-ha-worker3  Started container
  Warning  BackOff    2m55s (x252 over 57m)  kubelet, moelove-ha-worker3  Back-off restarting failed container

```

2, Go into the docker env for notebook pods and change permissions
```
docker exec -ti moelove-ha-worker3 /bin/bash
root@moelove-ha-worker3:/#chmod 777 /home/jovyan/
root@moelove-ha-worker3:/#
root@moelove-ha-worker3:/# ls -ltr /home/
total 0
drwxrwxrwx 2 root root 6 Jun 21 05:31 jovyan
root@moelove-ha-worker3:/#
```

**Comment by kavemun** [comment]
Hi @imicer 

No worries, I'm quite new myself and exploring. 
The permission you should change should be the folder on your node where your persistent volume is mapped to, not inside your container. 

First time you create your notebook, select new workspace and one will be generated for you but with permission 755. you can check by doing 

```
$ kubectl describe pv pvc-405f98fe-9266-11e9-a05d-5254004e297e
Name:            pvc-405f98fe-9266-11e9-a05d-5254004e297e
Labels:          <none>
Annotations:     cephFSProvisionerIdentity: cephfs-provisioner-1
                 cephShare: kubernetes-dynamic-pvc-45891efc-9266-11e9-a94f-22953f57fef7
                 pv.kubernetes.io/provisioned-by: ceph.com/cephfs
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    cephfs
Status:          Bound
Claim:           kubeflow/kubeflow-workspace
Reclaim Policy:  Delete
Access Modes:    RWO
VolumeMode:      Filesystem
Capacity:        2Gi
Node Affinity:   <none>
Message:         
Source:
    Type:        CephFS (a CephFS mount on the host that shares a pod's lifetime)
    Monitors:    [10.60.1.34:6789 10.60.1.35:6789 10.60.1.36:6789]
    Path:        /kubestorage/kubernetes/kubernetes-dynamic-pvc-45891efc-9266-11e9-a94f-22953f57fef7
    User:        kubernetes-dynamic-user-45891f99-9266-11e9-a94f-22953f57fef7
    SecretFile:  
    SecretRef:   &SecretReference{Name:ceph-kubernetes-dynamic-user-45891f99-9266-11e9-a94f-22953f57fef7-secret,Namespace:kubeflow,}
    ReadOnly:    false
Events:          <none>
```

if you look under "Path" you will find the directory where the persistent volume is mapped to.
you need to change the permission for that folder to 777.

After that you should be able to use that folder as your workspace and write to it.

**Comment by imicer** [comment]
Hi @kavemun 
Thank you for your kindly help. I will try your method in my env.

**Comment by imicer** [comment]
Hi @kavemun 
I tried to check the parameters for configuration in kubeflow UI .
I found the following things, I am not sure if it is design behaviors. 
How can I modify or create new pv for my case, Thanks.
(1) I can't specifiy the Mount Path for new notebook-workspaces though kubelow UI.

(2) when I used the default MountPath(/home/jovyan/) to create new Notebool server(nb001) in UI, I go to check PVC and PV

```
kubectl -n kubeflow get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                     STORAGECLASS   REASON   AGE
mypv                                       10Gi       RWX            Recycle          Bound    kubeflow/myclaim                                  22h
pvc-2e69f560-971a-11e9-9fbb-0242822b5788   10Gi       RWX            Delete           Bound    kubeflow/nb001            standard                20m
pvc-52240b06-9434-11e9-9270-0242a29cfe84   10Gi       RWO            Delete           Bound    kubeflow/katib-mysql      standard                3d16h
pvc-5ec8ecde-96ec-11e9-9fbb-0242822b5788   10Gi       RWO            Delete           Bound    kubeflow/data-vol-1       standard                5h48m
pvc-6a4da15e-9434-11e9-9270-0242a29cfe84   20Gi       RWO            Delete           Bound    kubeflow/mysql-pv-claim   standard                3d16h
pvc-6a8aa708-9434-11e9-9270-0242a29cfe84   20Gi       RWO            Delete           Bound    kubeflow/minio-pvc        standard                3d16h

kubectl -n kubeflow describe pv pvc-2e69f560-971a-11e9-9fbb-0242822b5788
Name:            pvc-2e69f560-971a-11e9-9fbb-0242822b5788
Labels:          <none>
Annotations:     kubernetes.io/createdby: hostpath-dynamic-provisioner
                 pv.kubernetes.io/bound-by-controller: yes
                 pv.kubernetes.io/provisioned-by: kubernetes.io/host-path
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    standard
Status:          Bound
Claim:           kubeflow/nb001
Reclaim Policy:  Delete
Access Modes:    RWX
VolumeMode:      Filesystem
Capacity:        10Gi
Node Affinity:   <none>
Message:
Source:
    Type:          HostPath (bare host directory volume)
    Path:          /tmp/hostpath_pv/2e6cb4a6-971a-11e9-96fc-0242a29cfe84
    HostPathType:
Events:            <none>

kubectl -n kubeflow describe pvc nb001
Name:          nb001
Namespace:     kubeflow
StorageClass:  standard
Status:        Bound
Volume:        pvc-2e69f560-971a-11e9-9fbb-0242822b5788
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/host-path
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWX
VolumeMode:    Filesystem
Events:
  Type       Reason                 Age   From                         Message
  ----       ------                 ----  ----                         -------
  Normal     ProvisioningSucceeded  22m   persistentvolume-controller  Successfully provisioned volume pvc-2e69f560-971a-11e9-9fbb-0242822b5788 using kubernetes.io/host-path
Mounted By:  nb001-0
```
BTW,
I have a test on creating the pod on NFS mount point. the pod can work well.

**Comment by imicer** [comment]
Hi
Now I can create notebook server, My steps are the following:
Firstly, I manually created pv and pvc, and then used Existing option in workspace volume of kubeflow UI. Please note that you should fill the pvc name created above to spawn.

**Comment by jbottum** [comment]
@kavemun did you get this resolved?  Have you tried with Kubeflow 0.6?  Did you try to set-up a persistent volume when your created the notebook in the data volume section?

**Comment by kavemun** [comment]
hi @jbottum,
yes It is mostly due to the way our persistent volume layer is setup, once i got my admin to sort it out, it worked, I was trying it out on version 0.6 too, no issues there
